content/blog/2022/2022-03-28-complete-guide-to-errors-in-express.md

```diff
@@ -0,0 +1,559 @@
+}
+
+// route handlers
+app.get(...)
+app.post(...)
+
+// attach error handling middleware functions after route handlers
+app.use(errorHandler)
+```
+These error-handling middleware functions are attached to the `app` instance after the route handler functions have been defined. 
+
+The built-in default error handler of Express described in the previous section is also an error-handling middleware function and is attached at the end of the middleware function stack if we do not define any error-handling middleware function.
+
+Any error in the route handlers gets propagated through the middleware stack and is handled by the last middleware function which can be the default error handler or one or more custom error-handling middleware functions if defined.
+
+## Calling the Error Handling Middleware Function
+
+When we get an error in the application, the error object is passed to the error-handling middleware, by calling the `next(error)` function as shown below:
+
+```js
+const express = require('express')
+const axios = require("axios")
+const app = express()
+
+const errorHandler = (error, request, response, next) {
+  // Error handling middleware functionality
+  console.log( `error ${error.message}`) // log the error
+  const status = error.status || 400
+  // send back an easily understandable error message to the caller
+  response.status(status).send(error.message)
+}
+
+app.get('/products', async (request, response)=>{
+  try{
+    const apiResponse = await axios.get("http://localhost:3001/products")
+
+    const jsonResponse = apiResponse.data
+    
+    response.send(jsonResponse)
+  }catch(error){
+    next(error) // calling next error handling middleware
+  }
+
+})
+app.use(errorHandler)
+```
+As we can see here, the `next(error)` function takes the error object in the `catch` block as input which is passed on to the next error-handling middleware function where we can potentially put the logic to extract relevant information from the error object, log the error, and send back an easily understandable error message to the caller.
+
+## Adding Multiple Middleware Functions for Error Handling 
+
+We can chain multiple error-handling middleware functions similar to what we do for other middleware functions.
+
+Let us define three middleware error handling functions and add them to our routes:
+
+```js
+
+// Error handling Middleware function for logging the error message
+const errorLogger = (error, request, response, next) => {
+  console.log( `error ${error.message}`) 
+  next(error) // calling next middleware
+}
+
+// Error handling Middleware function reads the error message 
+// and sends back a response in JSON format
+const errorResponder = (error, request, response, next) => {
+  response.header("Content-Type", 'application/json')
+    
+  const status = error.status || 400
+  response.status(status).send(error.message)
+}
+
+const invalidPathHandler = (request, response, next) => {
+  response.status(400)
```

Comment:
This should be `404`.

I'm missing the connection how this error handler is called when an unknown path is being called. Can you explain that in the text?